### DISCUSSION TITLE
How can i change form input value from header actions action closure?

### DISCUSSION THREAD

**Discussion by halil-cosdu-bear** [filamentphp/filament#10044] [Help] [answered]
Hello all,

how can i achieve change Form input value from action closure?

`class ViewTicket extends ViewRecord
{
    protected static string $resource = TicketResource::class;

    protected function getHeaderActions(): array
    {
        return [
            Actions\EditAction::make()->slideOver(),
            Action::make('reply')
                ->label('Reply Ticket')
                ->color($this->record->is_resolved ? 'danger' : 'success')
                ->disabled($this->record->is_resolved)->button()
                ->form([
                    RichEditor::make('message')
                        ->fileAttachmentsDisk('s3')
                        ->fileAttachmentsDirectory('attachments')
                        ->fileAttachmentsVisibility('private')
                        ->default($this->record->message)
                        ->required(),

         @leandrocfe           RichEditor::make('response') -> **_Want to change this input value from inside below action closure where todo is.
 
                ])
                ->action(function (Action $action, array $data) {
                
                 @leandrocfe         //TODO:: Change RichEditor value.
                          
                        $action->halt();
                    }
                })->slideOver()
        ];
    }
}`

**Comment by leandrocfe** [comment]
Sorry, I didn't understand what you need

**Reply by halil-cosdu-bear** [reply]
Hello thanks for reply,

In short story, In the above code you will see 

`                RichEditor::make('response') -> **_Want to change this input value_**
`
**I want to change this richeditor input value from** 

```
 ->action(function (Action $action, array $data) {
            
 **--> in here !!!**
                      
                      
                    $action->halt();
                }
```
            })->slideOver()

**Reply by leandrocfe** [reply]
`$data['response'] = 'xxx'`
?

**Reply by halil-cosdu-bear** [reply]
<img width="914" alt="Screenshot 2023-12-04 at 00 49 12" src="https://github.com/filamentphp/filament/assets/133899637/438b7eba-415e-4bf2-8dd5-41eae28884c5">

`protected function getHeaderActions(): array
    {
        return [
            Actions\EditAction::make()->slideOver(),
            Action::make('reply')
                ->label('Reply Ticket')
                ->color($this->record->is_resolved ? 'danger' : 'success')
                ->disabled($this->record->is_resolved)
                ->extraModalFooterActions(fn(Action $action): array => [
                    $action->makeModalSubmitAction('getsuggestion', arguments: ['halt' => true]),
                ])->button()
                ->form([
                    RichEditor::make('message')
                        ->fileAttachmentsDisk('s3')
                        ->fileAttachmentsDirectory('attachments')
                        ->fileAttachmentsVisibility('private')
                        ->default($this->record->message)
                        ->required(),

                    TextInput::make('suggest')
                        ->label('ChatGPT Directives')
                    ,

                    TextInput::make('response')
                ])
                ->action(function (Action $action, array $data) {
                    ;
                    if ($action->getModalAction('getsuggestion')->getArguments()['halt'] ?? false) {
                        $deviceInformationString = "My platform is : {$this->record->device['os_type']},
                            My platform version is : {$this->record->device['os_version']},
                            My device is : {$this->record->device['device_type']}";
                        $personalInformation = "My name is : {$this->record->name},
                            My email is : {$this->record->email}";
                        $question = "My device informations are: $deviceInformationString and my personal
                            informations are: $personalInformation. Use this information for better answering.
                            My question is: " . strip_tags($data['message']);
                        $chatGPT = \Http::chat()->post('/', ['question' => $question]);
                        $chatGPTResponse = $chatGPT->json()['text'];

                        $data['response'] = $chatGPTResponse; // nothing happens


                        $action->halt();
                    }
                  
                })->slideOver()
        ];
    }`

**Reply by halil-cosdu-bear** [reply]
$data['response'] = $chatGPTResponse; 

 $this->refreshFormData($data);
 
 this is also not working.

**Reply by leandrocfe** [reply, marked as answer]
Hum, ok. Use
```php
$this->mountedActionsData[0]['response'] = 'xx';
```

**Reply by halil-cosdu-bear** [reply]
@leandrocfe You are fantastic thanks for help !!!

**Reply by leandrocfe** [reply]
another way if you want
```php
$this->getMountedActionForm()->getComponents()[1]->state('xx') //0 = message, 1 = response
```

**Reply by halil-cosdu-bear** [reply]
Thanks for reply, god im still in here 🗡️ 

As you know we have 3 buttons,

1. Submit
2. Get Suggestion
3. Cancel

How i can detect "Get Suggestion" button triggered? 
I cant find a way detect which button clicked...

`$action->getModalAction('getsuggestion')->isTriggered()` ?? Maybe something like this. idk

**Reply by leandrocfe** [reply]
Did you try it using array $arguments?

**Reply by halil-cosdu-bear** [reply]
yes, its empty;

$action->getModalAction('getsuggestion')->getArguments()['halt'] ?? false) -> this is okay i can get halt variable but its useless because its always either true or false.

`->action(function (array $data, Action $action, array $arguments) {
                   //  dump($arguments); -> [] -> empty array
                    
                    if ($action->getModalAction('getsuggestion')->getArguments()['halt']) {
                    
                    // has to work only get suggestion triggered
                        $deviceInformation = $this->getDeviceInformation($data['message']);
                        $messages = $this->prepareGptMessages($deviceInformation);
                        $this->getMountedActionForm()->getComponents()[1]->disabled(); // this is also not working .))
                        $GptResponse = $this->sendGptRequest($messages);
                        $this->getMountedActionForm()->getComponents()[1]->disabled(false); // this is also not working .))
                        $structuredResponse = $this->formatGptResponse($GptResponse);
                        // $this->mountedActionsData[0]['response'] = $chatGPTResponse;
                        $this->getMountedActionForm()->getComponents()[1]->state($structuredResponse);
                        $action->halt();
                    }`       
                    // Submit button will work in here

**Reply by halil-cosdu-bear** [reply]
->extraModalFooterActions(fn(Action $action): array => [
                    $action->make('suggestion')
                        ->label('Get Suggestion')
                        ->action(Closure::fromCallable([$this, 'suggestion']))
                        ->button()
])
                
 BTW  found something like this : https://github.com/filamentphp/filament/issues/7324
           
           
```
public function suggestion(): void
    {
        $deviceInformation = $this->getDeviceInformation($this->mountedActionsData[0]['message']);
        $messages = $this->prepareGptMessages($deviceInformation);

        $GptResponse = $this->sendGptRequest($messages);
        $structuredResponse = $this->formatGptResponse($GptResponse);
        $this->mountedActionsData[0]['response'] = $structuredResponse;
    }
```

**Comment by rahulhaque** [comment]
In case someone stumble upon here with the same need, in Filament 5 we can also achieve this by -
```php
->action(function (array $data, Schema $schema, Action $action) {
  $schema->getComponentByStatePath('form_field_name')->state('value');
  $action->halt();
}
```
